### PR TITLE
Store series labels as a single string, to save memory

### PR DESCRIPTION
**Pull request by bboreham**
#### What this PR does

This is porting Mimir to use https://github.com/prometheus/prometheus/pull/10991

~Still a bit experimental.~

~Requires a couple more changes in mimir-prometheus; currently overriding to branch `fixup-labels-usage3`.~

#### Checklist

- [x] Tests updated
- NA Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by replay** [comment]
The CHANGELOG has just been cut to prepare for the next Mimir release. Please rebase `main` and eventually move the CHANGELOG entry added / updated in this PR to the top of the CHANGELOG document. Thanks!

File: pkg/util/metrics_helper.go [hunk 1]
```diff
@@ -714,18 +714,19 @@ func (r *UserRegistries) BuildMetricFamiliesPerUser() MetricFamiliesPerUser {
 }
 
 // FromLabelPairsToLabels converts dto.LabelPair into labels.Labels.
-func FromLabelPairsToLabels(pairs []*dto.LabelPair) labels.Labels {
-	builder := labels.NewBuilder(nil)
+func FromLabelPairsToLabels(builder *labels.SimpleBuilder, pairs []*dto.LabelPair) labels.Labels {
```
Comment (pracucci) [inline pkg/util/metrics_helper.go:717]: Labels in `pairs` are guaranteed to be always sorted, right?

**Review by pracucci — COMMENTED** [review]
Good job! There's quite a lot to digest here 😉  I took a first pass and overall LGTM. I left few comments here and there.

File: pkg/mimirpb/compat.go [hunk 2]
```diff
@@ -55,61 +55,30 @@ func ToWriteRequest(lbls []labels.Labels, samples []Sample, exemplars []*Exempla
 }
 
 // FromLabelAdaptersToLabels casts []LabelAdapter to labels.Labels.
-// It uses unsafe, but as LabelAdapter == labels.Label this should be safe.
-// This allows us to use labels.Labels directly in protos.
-//
-// Note: while resulting labels.Labels is supposedly sorted, this function
-// doesn't enforce that. If input is not sorted, output will be wrong.
+// For now it's doing an expensive conversion: TODO figure out a faster way, maybe via PreallocTimeSeries.
```
[the remark this answers is not in the document]
Comment (bboreham) [reply pkg/mimirpb/compat.go]: I made it faster on write path - see `FromLabelAdaptersOverwriteLabels`.
It's still more expensive than a cast, but showing around 3% of CPU in mimir-dev-09 right now.

I haven't looked at impact in read path yet.

**Comment by pracucci** [comment]
@bboreham Could you rebase this, please?

File: pkg/mimirpb/compat.go [hunk 3]
```diff
@@ -89,62 +89,51 @@ func (req *WriteRequest) AddHistogramSeries(lbls []labels.Labels, histograms []H
...
-// It uses unsafe, but as LabelAdapter == labels.Label this should be safe.
-// This allows us to use labels.Labels directly in protos.
+// For now it's doing an expensive conversion: TODO figure out a faster way.
 func FromLabelsToLabelAdapters(ls labels.Labels) []LabelAdapter {
-	return *(*[]LabelAdapter)(unsafe.Pointer(&ls))
+	r := make([]LabelAdapter, 0, ls.Len())
+	ls.Range(func(l labels.Label) {
+		r = append(r, LabelAdapter{Name: l.Name, Value: l.Value})
+	})
+	return r
```
Comment (dimitarvdimitrov) [inline pkg/mimirpb/compat.go:136]: This code seems to be used on the "hot" path. Not it seems to do an allocation per series for push requests from the ruler ([code](https://github.com/grafana/mimir/blob/bc241d994d70f78e42f8705f5ec83dda903e4026/pkg/ruler/compat.go#L87-L88)), for applying relabeling rules in the distributor ([code](https://github.com/grafana/mimir/blob/70a1cfc974b39fbfd8947d53ff9442f38eae401f/pkg/distributor/distributor.go#L786)), and once/twice for each touched timeseries by a query ([in the querier here](https://github.com/grafana/mimir/blob/1e262759d638eff986d3348a5fd4e3593f6cd39d/pkg/querier/remote_read.go#L257), [in the store-gateway here](https://github.com/grafana/mimir/blob/e16d1e397e91a43efe0c6ddb960a6ef804217961/pkg/storegateway/bucket.go#L664)). So in these cases we would hold the memory for the optimized stringlabels once and once for the unoptimized slice representation.

Have you ran any benchmarks to gauge the impact of this?

Comment (bboreham) [reply pkg/mimirpb/compat.go:136]: I have run benchmarks, but the underlying code keeps changing so it is hard to keep track.
I think it's basically whack-a-mole now; we have to run it again and see if anything pops out as a hotspot.

Relabelling is already an expensive operation with regexps; the code you pointed to is [different](https://github.com/grafana/mimir/blob/labels-as-string/pkg/distributor/distributor.go#L786) in my fork.

Comment (bboreham) [reply pkg/mimirpb/compat.go:136]: I'll make a start on running the Mimir benchmarks again.

Comment (dimitarvdimitrov) [reply pkg/mimirpb/compat.go:136]: another option is to run with stringlabels and without side-by-side for some time. Some ingester zones and some store-gateway zones can run different configurations. IIRC we can also do something similar for the queriers in Grafana Labs clusters.

Is it more difficult to get two concurrent container images with and without stringlabels?

**Comment by bboreham** [comment]
Slight change of tack: Mimir now builds with or without `-tags stringlabels`, and its efficiency is identical without `-tags stringlabels`. 

This is to allow us to build two images to ease rollback if a new issue is detected with `-tags stringlabels`.

**Review by colega — APPROVED** [review]
LGTM!

I would suggest adding a `test-with-stringlabels` job to the workflows, copying `test` and running it with stringlabels, so you won't need to chase down all the new usages of `nil` introduced by other PRs.

File: pkg/mimirpb/compat.go (hunk 3 above)
Comment (dimitarvdimitrov) [reply pkg/mimirpb/compat.go]: Do I read this right that it's allocating 96% more memory?
